src/core/qgsproject.h

```diff
@@ -575,6 +576,54 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
      */
     QMap<QString, QgsMapLayer *> mapLayers() const;
 
+    /**
+     * Unzip a project
+     * \param filename The zip file to unzip
+     * \returns true if unzip is well performed, false otherwise
+     * \since QGIS 3.0
+     */
+    bool unzip( const QString &filename );
+
+    /**
+     * Unzip a project with the current zip filename
+     * \returns true if unzip is well performed, false otherwise
+     * \since QGIS 3.0
+     */
+    bool unzip();
+
+    /**
+     * Zip the project
+     * \param filename The zip filename
+     * \returns true if zip is well performed, false otherwise
+     * \since QGIS 3.0
+     */
+    bool zip( const QString &filename );
+
+    /**
+     * Zip the project with the current zip filename
+     * \returns true if zip is well performed, false otherwise
+     * \since QGIS 3.0
+     */
+    bool zip();
+
+    /**
+     * Returns the current zip filename or an empty string if none.
+     * \since QGIS 3.0
+     */
+    QString zipFileName() const;
```

Comment:
Ok. I'm not sold on that approach though - I think it'll mean there's a lot of extra logic which needs to be scattered throughout the code to handle this check if zip file name, if so use it, if not, use filename. E.g. just of the top of my head the project properties window shows file name, and there's an expression variable for file name. There's probably other use of QgsProject.fileName() too. Any code which handles translation of paths relative to the project path will also need to be updated, because those relative paths will be to the location of the qgz file.

I'd suggest leaving QgsProject.fileName() and .fileInfo() as always returning paths for the original user-location of the project (regardless of whether it's a qgs or qgz file), and adding some other accessor which always returns the path to the original or extracted .qgs file.